instrumentation/dubbo/dubbo-apache-2.7/testing/src/main/groovy/io/opentelemetry/instrumentation/dubbo/apache/v2_7/AbstractDubboTest.groovy

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.dubbo.apache.v2_7
+
+import io.opentelemetry.instrumentation.dubbo.apache.v2_7.api.HelloService
+import io.opentelemetry.instrumentation.test.InstrumentationSpecification
+import io.opentelemetry.instrumentation.test.utils.PortUtils
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes
+import org.apache.dubbo.config.ApplicationConfig
+import org.apache.dubbo.config.ProtocolConfig
+import org.apache.dubbo.config.ReferenceConfig
+import org.apache.dubbo.config.ServiceConfig
+import org.apache.dubbo.config.bootstrap.DubboBootstrap
+import org.apache.dubbo.config.utils.ReferenceConfigCache
+import org.apache.dubbo.rpc.service.GenericService
+import spock.lang.Unroll
+
+import static io.opentelemetry.api.trace.Span.Kind.CLIENT
+import static io.opentelemetry.api.trace.Span.Kind.SERVER
+import static io.opentelemetry.instrumentation.test.utils.TraceUtils.basicSpan
+import static io.opentelemetry.instrumentation.test.utils.TraceUtils.runUnderTrace
+
+@Unroll
+abstract class AbstractDubboTest extends InstrumentationSpecification {
+
+  abstract ServiceConfig configureServer()
+
+  abstract ReferenceConfig<HelloService> configureClient(int port)
+
+  def "test apache dubbo base #dubbo"() {
+    setup:
+    def port = PortUtils.randomOpenPort()
+    def protocolConfig = new ProtocolConfig()
+    protocolConfig.setPort(port)
+
+    DubboBootstrap bootstrap = DubboBootstrap.getInstance()
+    bootstrap.application(new ApplicationConfig("dubbo-test-provider"))
+      .service(configureServer())
+      .protocol(protocolConfig)
+      .start()
+
+    def consumerProtocolConfig = new ProtocolConfig()
+    consumerProtocolConfig.setRegister(false)
+
+    def reference = configureClient(port)
+    DubboBootstrap consumerBootstrap = DubboBootstrap.getInstance()
+    consumerBootstrap.application(new ApplicationConfig("dubbo-demo-api-consumer"))
+      .reference(reference)
+      .protocol(consumerProtocolConfig)
+      .start()
+
+    when:
+    GenericService genericService = ReferenceConfigCache.getCache().get(reference) as GenericService
+    def o = new Object[1]
+    o[0] = "hello"
+    def response = runUnderTrace("parent") {
+      genericService.$invoke("hello", [String.getName()] as String[], o)
+    }
+
+    and:
+    def responseAsync = runUnderTrace("parent") {
+      genericService.$invokeAsync("hello", [String.getName()] as String[], o)
+    }
+
+    then:
+    response == "hello"
+    responseAsync.get() == "hello"
+
+    assertTraces(2) {
+      trace(0, 3) {
+        basicSpan(it, 0, "parent")
+        span(1) {
+          name "org.apache.dubbo.rpc.service.GenericService/\$invoke"
```

Comment:
I would generally expect the client service / method to match the server one. In dubbo, is it not possible for the client to have the name `HelloService/hello` instead of this generic name? I see `setInterface` and the method name `hello` in the invocation, so it seems like the client has all the information.

Reply:
Dubbo will use javaassist to set up a proxy for the interface to call directly instead of generic invoke, but I find that the proxy cannot work in groovy's unit test.So I use generic invoke here instead.

Reply:
I see - we'll need to add a smoke-test to verify the behavior in a future PR then. Can you add a comment here?

Reply:
I'd love to do it

Reply:
Ah sorry I meant a comment in the code :P `// Dubbo's codegen does not work properly from Groovy tests, so we need to verify in a smoke test instead.`

Reply:
Can you go ahead and add this code comment suggestion?